include/oneapi/dpl/pstl/hetero/dpcpp/unseq_backend_sycl.h

```diff
@@ -202,103 +202,132 @@ struct __init_processing
+            }
+        }
+        // _VecSize-wide vectorized path (__iters_per_work_item are multiples of _VecSize)
+        if (__is_full_wg)
+        {
+            vectorized_reduction_first(__adjusted_global_id, __res, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+                vectorized_reduction_remainder(__adjusted_global_id + __i * __stride, __res, __acc...);
         }
+        // At least one vector operation
+        else if (__adjusted_global_id + __vec_size_minus_one < __adjusted_n)
+        {
+            vectorized_reduction_first(__adjusted_global_id, __res, __acc...);
+            for (_Size __i = 1; __i < __no_vec_ops; ++__i)
+            {
+                const _Size __base_idx = __adjusted_global_id + __i * __stride;
+                if (__base_idx + __vec_size_minus_one < __adjusted_n)
+                    vectorized_reduction_remainder(__base_idx, __res, __acc...);
+                else if (__base_idx < __adjusted_n)
+                    scalar_reduction_remainder(__base_idx, __adjusted_n, __vec_size_minus_one, __res, __acc...);
+                else
+                    break;
+            }
+        }
+        // Scalar remainder
         else if (__adjusted_global_id < __adjusted_n)
         {
-            const _Size __items_to_process =
-                std::max(((__adjusted_n - __adjusted_global_id - 1) / __stride) + 1, static_cast<_Size>(0));
             new (&__res.__v) _Tp(__unary_op(__adjusted_global_id, __acc...));
-            for (_Size __i = 1; __i < __items_to_process; ++__i)
-                __res.__v = __binary_op(__res.__v, __unary_op(__adjusted_global_id + __stride * __i, __acc...));
+            scalar_reduction_remainder(static_cast<_Size>(__adjusted_global_id + 1), __adjusted_n,
+                                       static_cast<_Size>(_VecSize - 2), __res, __acc...);
         }
-        return;
-    }
-
-    // For non-SPIR-V targets, we check if the operator is commutative before selecting the appropriate codepath.
-    // On SPIR-V targets, the sequential implementation with the non-commutative operator is currently more performant.
-    static constexpr bool __use_nonseq_impl = !oneapi::dpl::__internal::__is_spirv_target_v && _Commutative::value;
-
-    template <typename _NDItemId, typename _Size, typename _Res, typename... _Acc>
-    inline void
-    operator()(const _NDItemId& __item_id, const _Size& __n, const _Size& __global_offset, _Res& __res,
-               const _Acc&... __acc) const
-    {
-        if constexpr (__use_nonseq_impl)
-            return nonseq_impl(__item_id, __n, __global_offset, __res, __acc...);
-        else
-            return seq_impl(__item_id, __n, __global_offset, __res, __acc...);
     }
 
     template <typename _Size>
     _Size
-    output_size(const _Size& __n, const ::std::uint16_t& __work_group_size) const
+    output_size(const _Size& __n, const _Size& __work_group_size, const _Size& __iters_per_work_item) const
```

Comment:
Small comment: should we really pass `_Size` data types by reference? May be simple by value?

Reply:
While this was not introduced in this PR, I agree that we should pass by value here and changed that.